crates/oci-distribution/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+#[macro_use]
+extern crate serde;
+
+use hyperx::header::Header;
+use oauth2::{Client as OauthClient, StandardToken};
+use serde::Deserialize;
+use std::convert::TryFrom;
+use url::Url;
+use www_authenticate::{Challenge, ChallengeFields, RawChallenge, WwwAuthenticate};
+
+use crate::errors::*;
+use crate::reference::Reference;
+
+const OCI_VERSION_KEY: &str = "Docker-Distribution-Api-Version";
+
+pub mod errors;
+pub mod reference;
+
+type OciResult<T> = Result<T, anyhow::Error>;
+
+struct Client {
+    token: Option<StandardToken>,
+}
+
+impl Default for Client {
+    fn default() -> Self {
+        Client { token: None }
+    }
+}
+
+impl Client {
+    /// According to the v2 specification, 200 and 401 error codes MUST return the
+    /// version. It appears that any other response code should be deemed non-v2.
+    ///
+    /// For this implementation, it will return v2 or an error result. If the error is a
+    /// `reqwest` error, the request itself failed. All other error messages mean that
+    /// v2 is not supported.
+    pub async fn version(&self, host: String) -> OciResult<String> {
+        let url = format!("https://{}/v2/", host);
+        let res = reqwest::get(&url).await?;
+        let disthdr = res.headers().get(OCI_VERSION_KEY);
+        let version = disthdr
+            .ok_or_else(|| anyhow::format_err!("no header v2 found"))?
+            .to_str()?
+            .to_owned();
+        Ok(version)
+    }
+
+    /// Perform an OAuth v2 auth request if necessary.
+    ///
+    /// This performs authorization and then stores the token internally to be used
+    /// on other requests.
+    pub async fn auth(&mut self, host: String, secret: Option<String>) -> OciResult<()> {
+        let cli = reqwest::Client::new();
+        // The version request will tell us where to go.
+        let url = format!("https://{}/v2/", host);
+        let res = cli.get(&url).send().await?;
+        let disthdr = res.headers().get(reqwest::header::WWW_AUTHENTICATE);
+        if disthdr.is_none() {
+            // The Authenticate header can be set to empty string.
+            return Ok(());
+        }
+        // FIXME: There must be a more elegant way of passing a header from hyper into WwwAuthenticate.
+        let auth = WwwAuthenticate::parse_header(&disthdr.unwrap().to_str()?.to_string().into())?;
+        let challenge_opt = auth.get::<BearerChallenge>();
+        if challenge_opt.is_none() {
```

Comment:
Same as above. One other thing you could do is:

`let challenge_opt = challenge_opt.ok_or(Ok(())?`